### PR TITLE
Smtp confirmable

### PR DESCRIPTION
**Pull request by Sany18**

### CODE REVIEW

File: app/mailers/device_mailer.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class DeviseMailer < Devise::Mailer
+  layout 'mailer'
+  default template_path: 'devise/mailer'
+  helper :application # gives access to all helpers defined within `application_helper`.
+  include Devise::Controllers::UrlHelpers # Optional. eg. `confirmation_url`
+
+  # def confirmation_instructions(record, token, opts={})
+  #   headers["Custom-header"] = "Confirm registration"
```
Comment (eugenekorpan) [inline app/mailers/device_mailer.rb]: never push commented out code into repository

File: db/seeds.rb [hunk 2]
```diff
@@ -5,3 +5,25 @@
 #
 #   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
 #   Character.create(name: 'Luke', movie: movies.first)
+include Faker
+
+for i in 0..16
+  Task.create(
+      id: Number.number( 5 ),
```
Comment (eugenekorpan) [inline db/seeds.rb]: id is automatically assigned by database

File: db/seeds.rb [hunk 3]
```diff
@@ -5,3 +5,25 @@
...
+      due_date: Date.today,
+      user_id: 1,
+      is_done: rand(2) == 1
+  )
+end
+
+User.create(
+    first_name: Name::first_name,
+    last_name: Name::last_name,
+    email: "w@w",
```
Comment (eugenekorpan) [inline db/seeds.rb]: use faker to generate random email

File: db/seeds.rb [hunk 4]
```diff
@@ -5,3 +5,25 @@
...
+include Faker
+
+for i in 0..16
+  Task.create(
+      id: Number.number( 5 ),
+      title: Book.title,
+      theme: Hacker.say_something_smart,
+      priority: rand( 1..5 ),
+      due_date: Date.today,
+      user_id: 1,
```
Comment (eugenekorpan) [inline db/seeds.rb]: id is automatically generated by database.
It is very possible that by this time there is not user with id 1 in the DB so you'll get an error.
Never use IDs directly (hardcoded)

File: db/seeds.rb [hunk 5]
```diff
@@ -5,3 +5,25 @@
...
+      is_done: rand(2) == 1
+  )
+end
+
+User.create(
+    first_name: Name::first_name,
+    last_name: Name::last_name,
+    email: "w@w",
+    password: "asdasd",
+    password_confirmation: "asdasd",
```
Comment (eugenekorpan) [inline db/seeds.rb]: Use good names for everything. EVERYTHING
please, remember it.

Comment (eugenekorpan) [reply db/seeds.rb]: your code describes you as a developer
bad names mean you are unreliable, irresponsible, unprofessional developer

File: db/seeds.rb (hunk 3 above)
Comment (Sany18) [reply db/seeds.rb]: цього юзера я використовую для тестування

File: db/seeds.rb (hunk 4 above)
Comment (Sany18) [reply db/seeds.rb]: тут та ж причина. id 1 для тестового юзера, який і створюється автоматично

File: app/mailers/device_mailer.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  #   super
+  # end
+  #
+  def confirmation_instructions(record, token, opts={})
+    if record.email.present?
+      opts[:subject] = "Welcome #{record.email.split('@').first.capitalize}, Confirm your account"
```
Comment (eugenekorpan) [inline app/mailers/device_mailer.rb]: https://github.com/Sany18/TaskManager/blob/master/app/models/user.rb#L11
user is required to enter an email, no need to check for it

File: db/migrate/20180923083128_create_tasks.rb [hunk 7]
```diff
@@ -4,7 +4,7 @@ def change
       t.string :title
       t.string :theme
       t.integer :priority
-      t.datetime :term
+      t.datetime :due_date
```
Comment (eugenekorpan) [inline db/migrate/20180923083128_create_tasks.rb]: what is it?

File: app/controllers/tasks_controller.rb [hunk 8]
```diff
@@ -66,7 +66,7 @@ def destroy
   # DELETE /task/delete_selected/:id
   def destroy_selected
     params[:id].split("&").map do |id|
-      current_user.tasks.find_by(id: id).destroy
+      Task.find_by(id: id).destroy
```
Comment (eugenekorpan) [inline app/controllers/tasks_controller.rb]: why have you switched from `current_user` to `Task`?
`current_user` was correct which meant to allow user to delete only tasks he owns. With new approach any user is able to delete any task. This is a big security issue

Comment (Sany18) [reply app/controllers/tasks_controller.rb]: i got error with last method:

Comment (Sany18) [reply app/controllers/tasks_controller.rb]: but destroy with current_user is work

File: spec/controllers/tasks_controller_spec.rb [hunk 9]
```diff
@@ -1,44 +1,56 @@
...
     end
   end
 
   # GET /task/:id
   context "#show" do
-    let(:task) { create(:task, user_id: @user.id) }
-    let(:params) {{ id: task.id }}
-    it "should show task" do
+    let!(:task) {create(:task, user_id: user.id)}
+    let!(:params) {{id: task.id}}
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]: не потрібно прям знак оклику ставити усім let

File: spec/controllers/tasks_controller_spec.rb [hunk 10]
```diff
@@ -47,64 +59,107 @@
...
-    let(:params) {{ id: task.id }}
-    it "should delete selected task" do
+    let!(:task) { create(:task, user_id: user.id) }
+    let!(:task_2) { create(:task, user_id: user.id) }
+    let!(:params) {{ id: [task.id, task_2.id] * '&' }}
+
+    it "destroy tasks" do
+      puts params
+      delete :destroy, params: params
+      expect( Task.find_by(id: task.id) ).to eq( nil )
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]: `..).to be_nil`
є куча чарівних матчерів

File: spec/controllers/tasks_controller_spec.rb (hunk 9 above)
Comment (Sany18) [reply spec/controllers/tasks_controller_spec.rb]: я впевнююсь що працюю зі створеним таском

File: app/controllers/tasks_controller.rb [hunk 11]
```diff
@@ -65,19 +65,18 @@ def destroy
...
-    params[:id].split("&").map do |id|
-      current_user.tasks.find_by(id: id).destroy
-    end
+    ids = params[:id].split("&")
+    current_user.tasks.where(id: ids).destroy_all
   end
 
   #GET /task/status_switch/:id
   def status_switch
     id = params[:id]
```
Comment (eugenekorpan) [inline app/controllers/tasks_controller.rb]: you don't  need to create additional local variable `id` here

File: app/controllers/tasks_controller.rb [hunk 12]
```diff
@@ -65,19 +65,18 @@ def destroy
...
-    task = Task.find_by(id: id)
-    if task.is_done == false
-      task.update(is_done: true)
+    @task = current_user.tasks.find_by(id: id)
+    if @task.is_done == false
```
Comment (eugenekorpan) [inline app/controllers/tasks_controller.rb]: you can simplify this condition.
you can change lines from 76 to 80 into just one simple line of code

File: spec/controllers/tasks_controller_spec.rb [hunk 13]
```diff
@@ -1,43 +1,56 @@
...
+    sign_in user
   end
 
   # GET /tasks
   context "#index" do
+    let!(:user_task) { create(:task, user_id: user.id) }
+    let!(:not_user_task) { create(:task, user_id: 0) }
+    it "have tasks" do
+      get :index
+      expect( assigns(:tasks).last ).to eq( user_task )
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]:
```
1 check the number of items in assigns(:tasks)
2 check that the only item in the assigns(:tasks) is your user_task
```

File: spec/controllers/tasks_controller_spec.rb [hunk 14]
```diff
@@ -1,43 +1,56 @@
...
-      expect(response).to have_http_status(302)
+      expect(response).to redirect_to( new_user_session_path )
     end
   end
 
   # GET /task/:id
   context "#show" do
-    let(:params) { task = create(:task); {id: task.id} }
-    it "should show task" do
+    let!(:task) {create(:task, user_id: user.id)}
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]: be consistent
https://github.com/rubocop-hq/ruby-style-guide

File: spec/controllers/tasks_controller_spec.rb (hunk 9 above)
Comment (eugenekorpan) [reply spec/controllers/tasks_controller_spec.rb]: на 34-у рядку ти викликаєш змінну `task`, в цей момент заносяться дані в базу, тому на 33-у тобу не потрібен знак оклику в `let`

File: spec/controllers/tasks_controller_spec.rb [hunk 15]
```diff
@@ -1,43 +1,56 @@
...
+    let!(:params) {{id: task.id}}
+
+    it "have task" do
+      get :show, params: params
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]: I think in this case you don't need to extract id into a separate `params`
just `get :show, id: task.id` should be fine

File: spec/controllers/tasks_controller_spec.rb [hunk 16]
```diff
@@ -46,59 +59,106 @@
 
   # GET /task/1/edit
   context "#edit" do
-    let(:params) { task = create(:task); {id: task.id} }
-    it "should direct to edit" do
+    let!(:task) {create(:task, user_id: user.id)}
+    let!(:params) {{id: task.id}}
+
+    it "have task" do
+      get :show, params: params
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]: ?
намагайся бути уважнішим. в тебе що контекст каже?
а ти куди реквести шлеш?

File: spec/controllers/tasks_controller_spec.rb [hunk 17]
```diff
@@ -46,59 +59,106 @@
...
       expect(response).to have_http_status(200)
     end
   end
 
   # POST /task
   context "#create" do
-    let(:params) { {task: build(:task).attributes} }
-    it "should create task" do
-      # sign_in FactoryBot.create(:user)
+    let!(:current_task) { build(:task) }
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]: тут не потрібен знак оклику

File: spec/controllers/tasks_controller_spec.rb [hunk 18]
```diff
@@ -46,59 +59,106 @@
...
+    let!(:params) {{ task: current_task.attributes }}
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]: тут також

File: spec/controllers/tasks_controller_spec.rb [hunk 19]
```diff
@@ -46,59 +59,106 @@
...
     end
   end
 
   # PATCH/PUT /task/:id
   context "#update" do
-    let(:params) { task = create(:task); {id: task.id, task: task.attributes} }
-    it "should update task" do
-      put :update, params: params
-      expect(response).to have_http_status(302)
+    let!(:task_1) { create(:task, user_id: user.id) }
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]: !

File: spec/controllers/tasks_controller_spec.rb [hunk 20]
```diff
@@ -46,59 +59,106 @@
...
+    let!(:params_1) {{ id: task_1.id, task: task_1.attributes }}
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]: !

Comment (eugenekorpan) [reply spec/controllers/tasks_controller_spec.rb]: you send the existing task attributes to check whether update is working. Sounds at least illogical
you need to send new attributes and check if back-end updated the attributes of the current task

File: spec/controllers/tasks_controller_spec.rb [hunk 21]
```diff
@@ -46,59 +59,106 @@
...
+
+    it "get current task" do
+      put :update, params: params_1
+      expect( assigns(:task) ).to eq( task_1 )
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]: this is update action, the main purpose of it is to update attributes of the task
but you actually don't check if attributes were updated so sounds like this test is useless

File: spec/controllers/tasks_controller_spec.rb [hunk 22]
```diff
@@ -46,59 +59,106 @@
...
   # PATCH/PUT /task/:id
   context "#update" do
-    let(:params) { task = create(:task); {id: task.id, task: task.attributes} }
-    it "should update task" do
-      put :update, params: params
-      expect(response).to have_http_status(302)
+    let!(:task_1) { create(:task, user_id: user.id) }
+    let!(:params_1) {{ id: task_1.id, task: task_1.attributes }}
+
+    it "get current task" do
```
Comment (eugenekorpan) [inline spec/controllers/tasks_controller_spec.rb]: Alright, below I can see you do check attributes, so this one is just redundant. Just delete it

File: spec/controllers/tasks_controller_spec.rb (hunk 14 above)
Comment (Sany18) [reply spec/controllers/tasks_controller_spec.rb]: after this task i will use gem robocop

File: spec/controllers/tasks_controller_spec.rb (hunk 15 above)
Comment (Sany18) [reply spec/controllers/tasks_controller_spec.rb]: i use params twice in current test

**Comment by eugenekorpan** [comment]
:+1:

**Pull request merged**